### DISCUSSION TITLE
messages remaining in noack

### DISCUSSION THREAD

**Discussion by FlaviuRadulescu** [php-amqplib/php-amqplib#1191] [Q&A] [answered]
In my PHP script, I have a loop that includes a non-blocking wait. If a message arrives during this, a callback function is triggered, and everything works as expected.

In this loop I also have code that "changes" the consumer's priority in the queue. To achieve this, I need to disconnect from the queue and reconnect with a higher priority because apparently is the only way to do it.

The issue I'm encountering is that between the time the wait finishes and the time I disconnect from the queue if the server pushes a message, that message remains in an "unacknowledged" state. This happens because the server delivers the message, but the callback is not executed since we're no longer in the wait state

We have multiple workers running on multiple servers, and sometimes certain resources need to be downloaded by a consumer to process a job. In these cases, we want the consumers with the local resources to be given higher priority, allowing them to pick up jobs faster than other consumers connected to the same queue. This process is dynamic and may change every 500 ms.

Any idea how I can solve this?

This is a pseudocode

```

    require_once __DIR__ . '/vendor/autoload.php';

    use PhpAmqpLib\Connection\AMQPStreamConnection;
    use PhpAmqpLib\Exception\AMQPTimeoutException;
    use PhpAmqpLib\Wire\AMQPTable;

    $connection = new AMQPStreamConnection('localhost', 5672, 'guest', 'guest');
    $channel    = $connection->channel();
    $channel->queue_declare('hello', false, false, false, false);
    $callback = function ($msg) {

        echo ' [x] Received ', $msg->body, "\n";
        $msg->ack(false);
    };
    $channel->basic_consume('hello', '', false, true, false, false, $callback);
    $consumerTag = "";
    $channel->basic_qos(null, 1, false);
    while ($channel->is_open()) {
        //some logic to change priority
        $changePriority = rand(0, 1);
        if ($changePriority and $consumerTag) {
            $channel->basic_cancel($consumerTag, false, true);
        }
        $arguments   = new AMQPTable(array(
                                         'x-priority'     => $changePriority ? 100 : 50,
                                         'x-max-priority' => 100,
                                     ));
        $consumerTag = $channel->basic_consume('hello', '', false, false, false, false, $callback);
        try {
            $channel->wait(null, true, 0);
        }
        catch (AMQPTimeoutException $e) {
        }
        usleep(100000);
    }
```

**Comment by lukebakken** [comment]
Hello, thanks for using this library and RabbitMQ.

What have _you_ tried to resolve this issue?

Here are some ideas off the top of my head:

* Use a blocking wait with a timeout. @ramunasd would know best, but my guess is this is preferable to `usleep`, anyway.
* After calling `basic_cancel`, close the channel and create a new one before calling `basic_consume`. This should re-queue any unacked messages

**Reply by FlaviuRadulescu** [reply]
Hi, thanks for getting back to me. I’ve been struggling with this issue for months, so I really appreciate your help.

Yes, I believe we initially set it up this way to have more precise control over the waiting time. If I remember correctly, the wait timeout is defined in seconds, but we wanted to check every 0.5 seconds. This setting can be easily adjusted, but I don’t think the issue is related to this.

We are currently subscribing to multiple queues on the same channel. My concern is that if we close the channel, it will disconnect all the queues. Would it be better if each queue subscription were handled on separate channels? Could this approach create issues in a setup with multiple channels?

**Reply by lukebakken** [reply]
> My concern is that if we close the channel, it will disconnect all the queues

Yes, it will. Use a channel per-`basic.consume`.

**Reply by lukebakken** [reply]
Really, the best way for us to assist you is for you to provide something we can clone and run to see the issue.

**Reply by FlaviuRadulescu** [reply]
> > My concern is that if we close the channel, it will disconnect all the queues
> 
> Yes, it will. Use a channel per-`basic.consume`.

Is there a limit or performance impact if we have many channels open instead of one?

**Reply by FlaviuRadulescu** [reply]
> Really, the best way for us to assist you is for you to provide something we can clone and run to see the issue.

We will try to provide a code. The problem we have is that is hard to replicate, but is happening several times a day. From what we have investigated it seems that we are closing a connection when a message is already in some kind of a queue waiting for the wait to pick it up, at this stage we are not sure if the acknowledge doesn't work. or if the wait will not pick up the message anymore because we have reconnected to the queue (on the same channel)

**Reply by lukebakken** [reply]
> Is there a limit or performance impact if we have many channels open instead of one?

Only you can determine that in your environment.

> it seems that we are closing a connection when a message is already in some kind of a queue

That's probably not what is happening. If a connection is closed with un-acked messages, RabbitMQ will return those messages to the Ready state, because all channels associated with that connection will have been closed as well.

**Reply by lukebakken** [reply]
Right now, don't worry about providing code. I've given you some ideas for how to experiment and try to fix this issue.

**Reply by FlaviuRadulescu** [reply, marked as answer]
I just discovered that if I reuse the same consumer tag when I reconnect the problem disappears. Don't know the cause or the reason is working, but thank God it is working
